File: java/graph/shortest_path/BellmanFord.java
package graph.shortest_path;

import java.util.Arrays;

public class BellmanFord {
    public int[] search(VertexNode[] graphs, int start) {
        int[] result = new int[graphs.length];
        Arrays.fill(result, Integer.MAX_VALUE);
        result[start] = 0;

        int index = start;
        for (VertexNode _ : graphs) {
            EdgeNode edge = graphs[index].next;
            while (null != edge) {
                int pivot = result[index] + edge.weight;
                int edgeIndex = edge.index;
                if (Integer.MAX_VALUE != result[index] && pivot < result[edgeIndex]) {
                    result[edgeIndex] = pivot;
                }
                edge = edge.next;
            }

            ++index;
            if (index >= graphs.length) {
                index = 0;
            }
        }

        index = start;
        for (VertexNode _ : graphs) {
            EdgeNode edge = graphs[index].next;
            while (null != edge) {
                int pivot = result[index] + edge.weight;
                int edgeIndex = edge.index;
                if (Integer.MAX_VALUE != result[index] && pivot < result[edgeIndex]) {
                    return new int[0];
                }
                edge = edge.next;
            }

            ++index;
            if (index >= graphs.length) {
                index = 0;
            }
        }
        return result;
    }

    public class EdgeNode {
        public int index = 0;
        public int weight = 0;
        public EdgeNode next = null;

        public EdgeNode(int index, int weight) {
            this.index = index;
            this.weight = weight;
        }
    }

    public class VertexNode {
        public int value = 0;
        public EdgeNode next = null;

        public VertexNode(int value) {
            this.value = value;
        }
    }
}

File: cpp/graph/shortest_path/BellmanFord.cpp
#include <iostream>
#include <vector>
using namespace std;

struct EdgeNode {
    int index;
    int weight;
    EdgeNode *next;

    EdgeNode(int index, int weight) : index(index), weight(weight), next(nullptr) {}
};

struct VertexNode {
    int value;
    EdgeNode *next;

    VertexNode(int value) : value(value), next(nullptr) {}
};

class BellmanFord {
public:
    vector<int> search(vector<VertexNode *> &graphs, int start) {
        vector<int> result(graphs.size(), INT_MAX);
        result[start] = 0;

        int index = start;
        for (VertexNode *_ : graphs) {
            EdgeNode *edge = graphs[index]->next;
            while (edge) {
                int pivot = result[index] + edge->weight;
                int edgeIndex = edge->index;
                if (INT_MAX != result[index] && pivot < result[edgeIndex]) {
                    result[edgeIndex] = pivot;
                }
                edge = edge->next;
            }

            ++index;
            if (index >= (int) graphs.size()) {
                index = 0;
            }
        }

        index = start;
        for (VertexNode *_ : graphs) {
            EdgeNode *edge = graphs[index]->next;
            while (edge) {
                int pivot = result[index] + edge->weight;
                int edgeIndex = edge->index;
                if (INT_MAX != result[index] && pivot < result[edgeIndex]) {
                    return vector<int>(0);
                }
                edge = edge->next;
            }

            ++index;
            if (index >= (int) graphs.size()) {
                index = 0;
            }
        }
        return result;
    }
};
